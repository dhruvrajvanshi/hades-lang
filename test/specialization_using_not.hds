struct Box[T] {}
trait Foo[T] {
    def foo(): Void
}
implementation Foo[Bool] {
    def foo(): Void {}
}

implementation[T] Foo[Box[T]] where Foo[T] {
    def foo(): Void {
        puts(b"Contains Foo")
    }
}
implementation[T] Foo[Box[T]] where not Foo[T] {
    def foo(): Void {
        puts(b"Does not contain Foo")
    }
}

def foo[T](): Void where Foo[T] {
    Foo[T].foo()
}
def main(): Void {
    foo[Box[Bool]]()
    foo[Box[usize]]()
}


extern def puts(*u8): Void = puts
