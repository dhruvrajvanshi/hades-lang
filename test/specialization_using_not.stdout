Contains Foo
Does not contain Foo